_posts/gRPC-Go performance improvements.md

```diff
@@ -0,0 +1,148 @@
+
+
+### Piggyback window updates
+
+[Code link](https://github.com/grpc/grpc-go/pull/1273)
+
+Sending a window update itself has a cost associated to it; a flush operation is necessary, which results in a syscall. Syscalls are blocking and slow. Therefore, when sending out a stream-level window update, it makes sense to also check if a connection-level window update can be sent using the same flush syscall.
+
+
+### BDP estimation and dynamic flow control window
+
+[Code link](https://github.com/grpc/grpc-go/pull/1310)
+
+This feature  is the latest and in some ways the most awaited optimization feature that has helped us close the final gap between gRPC and HTTP/1.1 performance on high latency networks.
+
+Bandwidth Delay Product (BDP) is the bandwidth of a network connection times its round-trip latency.  This effectively tells us how many bytes can be "on the wire" at a given moment, if full utilization is achieved.
+
+The algorithm to compute BDP and adapt accordingly was first proposed by @ejona and later implemented by both gRPC-C core and gRPC-Java (note that it isn't enabled in Java yet). The idea is simple and powerful: every time a receiver gets a data frame it sends out a BDP ping (a ping with unique data only used by BDP estimator). After this, the receiver starts counting the number of bytes it receives (including the ones that triggered the BDP ping) until it receives the ack for that ping. This total sum of all bytes received in about 1.5 RTT (Round-Trip Time) is an approximation of the effective BDP * 1.5. If this is close to our current window size (say, more than 2/3rd of it) we must increase the window size. We put our window sizes (both streaming and connection) to be twice the BDP we sampled.
+
+An interesting thing to note here is that just this algorithm by itself could cause the BDP estimation to increase indefinitely; an increase in window will result in sampling more bytes which in turn will cause the window to be increased further. This phenomenon is called buffer-bloat and was discovered by earlier implementations in gRPC-C core and gRPC-Java. The solution to this is to calculate the bandwidth for every sample and check if it is greater than the maximum bandwidth noted so far. If so, only then increase our window sizes. The bandwidth, as we know, can be calculated by dividing the sample by RTT * 1.5 (remember the sample was for one and a half round trips). If the bandwidth doesn't increase with an increase in sampled bytes that's indicative that this change is because of an increased window size and doesn't really reflect the nature of the network itself. 
```

Comment:
"An interesting thing to note here is that" <== verbose 
The solution is to only increase window size when the calculated bandwidth is greater than the maximum observed bandwidth thus far. The bandwidth can be calculated by dividing the sample size by RTT*1.5.. 

Link to wikipedia on bufferbloat? https://en.wikipedia.org/wiki/Bufferbloat ?